src/mlpack/methods/lmnn/lmnn_main.cpp

```diff
@@ -0,0 +1,350 @@
+    "gradients. It primarily on six parameters: the step size (specified "
+    "with " + PRINT_PARAM_STRING("step_size") + "), the batch size (specified "
+    "with " + PRINT_PARAM_STRING("batch_size") + "), the maximum number of "
+    "passes (specified with " + PRINT_PARAM_STRING("passes") + ") "
+    ", the exponential decay rate for the first moment estimates (specified "
+    "with " + PRINT_PARAM_STRING("beta1") + "), the exponential decay rate for "
+    "weighted infinity norm (specified with " + PRINT_PARAM_STRING("beta2") +
+     "), epsilon (specified with " + PRINT_PARAM_STRING("epsilon") + ").  In "
+    "addition, a normalized starting point can be used by specifying the " +
+    PRINT_PARAM_STRING("normalize") + " parameter. "
+    "\n\n"
+    "BigBatch_SGD, specified by the value 'bbsgd' for the parameter " +
+    PRINT_PARAM_STRING("optimizer") + ", depends primarily on four parameters: "
+    "the step size (specified with " + PRINT_PARAM_STRING("step_size") + "), "
+    "the batch size (specified with " + PRINT_PARAM_STRING("batch_size") + "), "
+    "the maximum number of passes (specified with " +
+    PRINT_PARAM_STRING("passes") + "), the batch delta (specified "
+    "with " + PRINT_PARAM_STRING("batch_delta") + ").  In addition, "
+    "a normalized starting point can be used by specifying the " +
+    PRINT_PARAM_STRING("normalize") + " parameter. "
+    "\n\n"
+    "Stochastic gradient descent, specified by the value 'sgd' for the "
+    "parameter " + PRINT_PARAM_STRING("optimizer") + ", depends "
+    "primarily on three parameters: the step size (specified with " +
+    PRINT_PARAM_STRING("step_size") + "), the batch size (specified with " +
+    PRINT_PARAM_STRING("batch_size") + "), and the maximum number of passes "
+    "(specified with " + PRINT_PARAM_STRING("passes") + ").  In "
+    "addition, a normalized starting point can be used by specifying the " +
+    PRINT_PARAM_STRING("normalize") + " parameter. "
+    "\n\n"
+    "The L-BFGS optimizer, specified by the value 'lbfgs' for the parameter " +
+    PRINT_PARAM_STRING("optimizer") + ", uses a back-tracking line search "
+    "algorithm to minimize a function.  The following parameters are used by "
+    "L-BFGS: " + PRINT_PARAM_STRING("num_basis") + " (specifies the number"
+    " of memory points used by L-BFGS), " +
+    PRINT_PARAM_STRING("max_iterations") + ", " +
+    PRINT_PARAM_STRING("armijo_constant") + ", " +
+    PRINT_PARAM_STRING("wolfe") + ", " + PRINT_PARAM_STRING("tolerance") +
+    " (the optimization is terminated when the gradient norm is below this "
+    "value), " + PRINT_PARAM_STRING("max_line_search_trials") + ", " +
+    PRINT_PARAM_STRING("min_step") + ", and " +
+    PRINT_PARAM_STRING("max_step") + " (which both refer to the line search "
+    "routine).  For more details on the L-BFGS optimizer, consult either the "
+    "mlpack L-BFGS documentation (in lbfgs.hpp) or the vast set of published "
+    "literature on L-BFGS."
```

Comment:
This is great, thank you for having such clear and comprehensive documentation.  Could you add a bit about the `range`, `regularization`, and `normalization` parameters?  I think it would be nice to add an example call or two, do you think you could do that?  If you think that makes it too long, we can probably make the optimizer bits a little shorter, but I have no issue with long documentation.

Reply:
Done! Added examples too, Hope they are up to the expectations.